### PR TITLE
[JENKINS-23805] Add support for the security realm's provided case sensitivity

### PR DESCRIPTION
**Pull request by stephenc**
[JENKINS-23805](https://issues.jenkins-ci.org/browse/JENKINS-23805)

@reviewbybees

(this was sitting in my workspace for who knows how long)

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -1,72 +1,77 @@
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
```
Comment (jglick) [inline pom.xml]: Please revert unnecessary changes.

File: src/main/java/hudson/security/AuthorizationMatrixProperty.java [hunk 2]
```diff
@@ -238,10 +241,47 @@ public boolean isBlocksInheritance() {
...
+	 */
+	public boolean hasPermission(String sid, Permission p, boolean principal) {
+        final SecurityRealm securityRealm = Jenkins.getInstance().getSecurityRealm();
+        final IdStrategy strategy = principal ? securityRealm.getUserIdStrategy() : securityRealm.getGroupIdStrategy();
+		for (; p != null; p = p.impliedBy) {
+            if (!p.getEnabled()) {
+                continue;
+            }
+            Set<String> set = grantedPermissions.get(p);
+			if (set != null && set.contains(sid))
```
Comment (jglick) [inline src/main/java/hudson/security/AuthorizationMatrixProperty.java]: TABs

File: src/main/java/hudson/security/AuthorizationMatrixProperty.java [hunk 3]
```diff
@@ -238,10 +241,47 @@ public boolean isBlocksInheritance() {
...
+				return true;
+            if (set != null) {
```
Comment (jglick) [inline src/main/java/hudson/security/AuthorizationMatrixProperty.java]: Could collapse two checks for `set != null`.

**Comment by jglick** [comment]
:bee: though without test coverage this seems like a potentially dangerous change.

File: src/main/java/hudson/security/AuthorizationMatrixProperty.java [hunk 4]
```diff
@@ -238,10 +241,47 @@ public boolean isBlocksInheritance() {
 	 * Checks if the given SID has the given permission.
 	 */
 	public boolean hasPermission(String sid, Permission p) {
+        final SecurityRealm securityRealm = Jenkins.getInstance().getSecurityRealm();
```
Comment (amuniz) [inline src/main/java/hudson/security/AuthorizationMatrixProperty.java]: Null check required?

File: src/main/java/hudson/security/AuthorizationMatrixProperty.java [hunk 5]
```diff
@@ -238,10 +241,47 @@ public boolean isBlocksInheritance() {
...
+            }
+		}
+		return false;
+	}
+
+	/**
+	 * Checks if the given SID has the given permission.
+	 */
+	public boolean hasPermission(String sid, Permission p, boolean principal) {
+        final SecurityRealm securityRealm = Jenkins.getInstance().getSecurityRealm();
```
Comment (amuniz) [inline src/main/java/hudson/security/AuthorizationMatrixProperty.java]: Null check again.

File: src/main/java/hudson/security/AuthorizationMatrixProperty.java [hunk 6]
```diff
@@ -251,7 +291,19 @@ public boolean hasPermission(String sid, Permission p) {
      */
     public boolean hasExplicitPermission(String sid, Permission p) {
         Set<String> set = grantedPermissions.get(p);
-        return set != null && set.contains(sid);
+        if (set != null && p.getEnabled()) {
+            if (set.contains(sid))
+                return true;
+            final SecurityRealm securityRealm = Jenkins.getInstance().getSecurityRealm();
```
Comment (amuniz) [inline src/main/java/hudson/security/AuthorizationMatrixProperty.java]: Null check.

**Comment by amuniz** [comment]
:bee:

**Comment by stephenc** [comment]
@reviewbybees done

**Comment by jglick** [comment]
Needs merge with `master`; POM change should be a single line.

**Comment by RemcoTukker** [comment]
@jglick Could you find some time to merge and ideally also release this? It would resolve a longstanding and really annoying usability issue that affects a lot of people ( https://issues.jenkins-ci.org/browse/JENKINS-23805 and https://issues.jenkins-ci.org/browse/JENKINS-23844 ). Thanks!

**Comment by stephenc** [comment]
@reviewbybees refresh

File: pom.xml [hunk 7]
```diff
@@ -37,7 +37,7 @@
         <dependency> <!-- TODO JENKINS-33095 workaround -->
             <groupId>org.jenkins-ci.plugins.icon-shim</groupId>
             <artifactId>icon-shim</artifactId>
-            <version>2.0.3</version> 
+            <version>2.0.3</version>
```
Comment (jglick) [inline pom.xml]: Why?

File: src/main/java/hudson/security/ProjectMatrixAuthorizationStrategy.java [hunk 8]
```diff
@@ -138,7 +138,7 @@ public String getDisplayName() {
         private RobustReflectionConverter ref;
 
         public ConverterImpl(Mapper m) {
-            ref = new RobustReflectionConverter(m,new JVM().bestReflectionProvider());
+            ref = new RobustReflectionConverter(m,JVM.newReflectionProvider());
```
Comment (jglick) [inline src/main/java/hudson/security/ProjectMatrixAuthorizationStrategy.java]: Unrelated?

Comment (stephenc) [inline src/main/java/hudson/security/ProjectMatrixAuthorizationStrategy.java]: It's equivalent, but the preferred method call for XStream 1.4.5+

**Comment by stephenc** [comment]
@jglick can you cut a release please

**Comment by jglick** [comment]
done

**Comment by daniel-beck** [comment]
Appears to have caused https://issues.jenkins-ci.org/browse/JENKINS-46190

**Comment by daniel-beck** [comment]
This appears to not cover the folder property.
